File: test-security-rules.dart
import 'dart:io';
import 'dart:convert';

/// Test script for Firebase Security Rules
/// 
/// This script validates that the Firestore security rules work correctly
/// by testing various scenarios with different user roles and permissions.
/// 
/// Requirements tested:
/// - 5.1: User approval workflow
/// - 5.2: Admin approval functionality  
/// - 5.3: Admin role checks
/// - 5.4: Access control based on approval status
/// - 6.4: Data access security
void main() async {
  print('🔒 Firebase Security Rules Test Suite');
  print('=====================================\n');

  // Test scenarios
  final testResults = <String, bool>{};

  try {
    // Test 1: Unauthenticated access should be denied
    print('Test 1: Unauthenticated access...');
    testResults['unauthenticated_access'] = await testUnauthenticatedAccess();

    // Test 2: Unapproved user access should be limited
    print('Test 2: Unapproved user access...');
    testResults['unapproved_user_access'] = await testUnapprovedUserAccess();

    // Test 3: Approved leader access should work
    print('Test 3: Approved leader access...');
    testResults['approved_leader_access'] = await testApprovedLeaderAccess();

    // Test 4: Admin access should have full permissions
    print('Test 4: Admin access...');
    testResults['admin_access'] = await testAdminAccess();

    // Test 5: User can only access their own data
    print('Test 5: User data isolation...');
    testResults['user_data_isolation'] = await testUserDataIsolation();

    // Test 6: Version control for conflict resolution
    print('Test 6: Version control...');
    testResults['version_control'] = await testVersionControl();

    // Test 7: Data validation rules
    print('Test 7: Data validation...');
    testResults['data_validation'] = await testDataValidation();

    // Print results
    print('\n📊 Test Results Summary');
    print('=======================');
    
    int passed = 0;
    int total = testResults.length;
    
    testResults.forEach((test, result) {
      final status = result ? '✅ PASS' : '❌ FAIL';
      print('$status $test');
      if (result) passed++;
    });
    
    print('\nOverall: $passed/$total tests passed');
    
    if (passed == total) {
      print('🎉 All security rules tests passed!');
      exit(0);
    } else {
      print('⚠️  Some security rules tests failed. Please review the rules.');
      exit(1);
    }
    
  } catch (e) {
    print('❌ Error running security tests: $e');
    print('\nMake sure the Firebase Emulator is running:');
    print('firebase emulators:start --only firestore,auth');
    exit(1);
  }
}

/// Test that unauthenticated users cannot access any data
Future<bool> testUnauthenticatedAccess() async {
  try {
    // Simulate unauthenticated request to users collection
    final result = await simulateFirestoreRequest(
      collection: 'users',
      operation: 'read',
      authenticated: false,
    );
    
    // Should be denied
    return result['allowed'] == false;
  } catch (e) {
    print('  Error: $e');
    return false;
  }
}

/// Test that unapproved users have limited access
Future<bool> testUnapprovedUserAccess() async {
  try {
    // Simulate unapproved user trying to access attendees
    final result = await simulateFirestoreRequest(
      collection: 'attendees',
      operation: 'read',
      authenticated: true,
      userId: 'unapproved-user',
      userRole: 'leader',
      userApproved: false,
    );
    
    // Should be denied
    return result['allowed'] == false;
  } catch (e) {
    print('  Error: $e');
    return false;
  }
}

/// Test that approved leaders can access and modify data
Future<bool> testApprovedLeaderAccess() async {
  try {
    // Test read access
    final readResult = await simulateFirestoreRequest(
      collection: 'attendees',
      operation: 'read',
      authenticated: true,
      userId: 'approved-leader',
      userRole: 'leader',
      userApproved: true,
    );
    
    // Test create access
    final createResult = await simulateFirestoreRequest(
      collection: 'attendees',
      operation: 'create',
      authenticated: true,
      userId: 'approved-leader',
      userRole: 'leader',
      userApproved: true,
      data: {
        'name': 'Test Attendee',
        'phoneNumber': '+1234567890',
        'location': 'Test Location',
        'category': 'student',
        'serviceId': 1,
        'registeredAt': DateTime.now().toIso8601String(),
        'createdBy': 'approved-leader',
        'createdAt': DateTime.now().toIso8601String(),
        'version': 1,
      },
    );
    
    // Both should be allowed
    return readResult['allowed'] == true && createResult['allowed'] == true;
  } catch (e) {
    print('  Error: $e');
    return false;
  }
}

/// Test that admins have full access
Future<bool> testAdminAccess() async {
  try {
    // Test user management (admin-only)
    final userUpdateResult = await simulateFirestoreRequest(
      collection: 'users',
      operation: 'update',
      authenticated: true,
      userId: 'admin-user',
      userRole: 'admin',
      userApproved: true,
      documentId: 'other-user',
      data: {
        'isApproved': true,
      },
    );
    
    // Test delete access (admin-only for message logs)
    final deleteResult = await simulateFirestoreRequest(
      collection: 'messageLogs',
      operation: 'delete',
      authenticated: true,
      userId: 'admin-user',
      userRole: 'admin',
      user

File: test-security-rules.js
// Firestore Security Rules Test Suite
// Run with: firebase emulators:exec --only firestore "node test-security-rules.js"

const firebase = require('@firebase/testing');
const fs = require('fs');

const PROJECT_ID = 'tuk-cu-messaging-test';
const RULES_FILE = './firestore.rules';

// Test data
const mockUsers = {
  admin: { uid: 'admin-user', role: 'admin', isApproved: true, email: 'admin@tuk.ac.ke' },
  leader: { uid: 'leader-user', role: 'leader', isApproved: true, email: 'leader@tuk.ac.ke' },
  member: { uid: 'member-user', role: 'member', isApproved: true, email: 'member@tuk.ac.ke' },
  unapproved: { uid: 'unapproved-user', role: 'leader', isApproved: false, email: 'pending@tuk.ac.ke' },
  unauthorized: null
};

const mockAttendee = {
  name: 'John Doe',
  phoneNumber: '+254712345678',
  location: 'Nairobi',
  category: 'student',
  serviceId: 1,
  registeredAt: firebase.firestore.Timestamp.now(),
  createdBy: 'leader-user',
  createdAt: firebase.firestore.Timestamp.now(),
  version: 1
};

const mockMessageLog = {
  attendeeId: 'attendee-123',
  message: 'Welcome to TUK CU!',
  sendStatus: 'delivered',
  sentAt: firebase.firestore.Timestamp.now(),
  serviceId: 1,
  sentBy: 'leader-user',
  createdAt: firebase.firestore.Timestamp.now(),
  version: 1
};

// Helper functions
function getFirestore(auth) {
  return firebase.initializeTestApp({
    projectId: PROJECT_ID,
    auth: auth
  }).firestore();
}

function getAdminFirestore() {
  return firebase.initializeAdminApp({
    projectId: PROJECT_ID
  }).firestore();
}

async function setupTestData() {
  const admin = getAdminFirestore();
  
  // Create test users
  for (const [key, user] of Object.entries(mockUsers)) {
    if (user) {
      await admin.collection('users').doc(user.uid).set({
        uid: user.uid,
        email: user.email,
        name: `Test ${key}`,
        role: user.role,
        isApproved: user.isApproved,
        createdAt: firebase.firestore.Timestamp.now()
      });
    }
  }
  
  console.log('✅ Test data setup complete');
}

async function runTests() {
  console.log('🔥 Starting Firestore Security Rules Tests');
  console.log('==========================================');
  
  let passed = 0;
  let failed = 0;
  
  // Test helper function
  async function test(description, testFn) {
    try {
      await testFn();
      console.log(`✅ ${description}`);
      passed++;
    } catch (error) {
      console.log(`❌ ${description}`);
      console.log(`   Error: ${error.message}`);
      failed++;
    }
  }
  
  // Authentication Tests
  console.log('\n📋 Authentication Tests');
  console.log('------------------------');
  
  await test('Unauthenticated users cannot read users collection', async () => {
    const db = getFirestore(null);
    const doc = db.collection('users').doc('admin-user');
    await firebase.assertFails(doc.get());
  });
  
  await test('Authenticated users can read their own profile', async () => {
    const db = getFirestore(mockUsers.leader);
    const doc = db.collection('users').doc('leader-user');
    await firebase.assertSucceeds(doc.get());
  });
  
  await test('Non-admin users cannot read other profiles', async 